### PR TITLE
Feature/#93 front end tests

### PR DESCRIPTION
**Pull request by leandertolksdorf**
Current test coverage:

Bildschirm­foto 2023-01-30 um 16 25 59

When running cypress tests, the coverage report gets saved into `coverage`folder, we could include this in the CI output.

I added more tests for frontend-only stuff like tooltips, but I'm really struggling with tests that involve signals from the backend. See file `cypress/e2e/2-indicators/line-indicators.cy.js`: I tried mocking the websocket connection to send a test sample array. And the message is received at the frontend. Problems are:

- The message data is received as a string like `1, 0, 0, ...` instead of an array buffer.
- Despite turning on the oscillator by clicking the start button, $osciEnabled is still false and the onmessage handler doesn't get to updating the other components.

### CODE REVIEW

**Review by PhlppKrmr — CHANGES_REQUESTED** [review]
Thanks for adding all those new tests :)

File: Apps/frontend/cypress/e2e/2-indicators/line-indicators.cy.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        cy.stub(window, "WebSocket").returns(mockSocket);
+      },
+    });
+  });
+
+  it("is visible", () => {
+    cy.get('[data-cy="line-indicators"]').should("be.visible");
+  });
+
+  it("updates min value", () => {
```
[the remark this answers is not in the document]
Comment (leandertolksdorf) [reply Apps/frontend/cypress/e2e/2-indicators/line-indicators.cy.js:25]: That's what i pointed out in the pr description. Unfortunately after many hours, I coudn't figure out how to fix it so I'm leaving it out for now.

File: Apps/frontend/src/components/GNDButton.svelte [hunk 2]
```diff
@@ -21,4 +22,4 @@
   }}
   data-cy="gnd-button"
 />
-<Tooltip target={button} placement="bottom">{TOOLTIP_BUTTON_GND}</Tooltip>
+<Tooltip target={button} placement="bottom" data-cy="gnd-button-tooltip">{TOOLTIP_BUTTON_GND}</Tooltip>
```
[the remark this answers is not in the document]
Comment (leandertolksdorf) [reply Apps/frontend/src/components/GNDButton.svelte:25]: How does it reformat? Doesn't do anything for me.

File: Apps/frontend/cypress/e2e/3-control-panel-top/gnd-button.cy.js [hunk 3]
```diff
@@ -13,4 +13,17 @@ describe("gndButton", () => {
   it("is clickable", () => {
     cy.get('[data-cy="gnd-button"]').click();
   });
+
+  it("shows tooltip on hover", () => {
+    cy.get('[data-cy="gnd-button"]').trigger("mouseover");
+    cy.get('[data-cy="gnd-button-tooltip"]').should("be.visible");
+  });
+  // Stub is not working in the e2e test, createEventDispatcher still
```
[the remark this answers is not in the document]
Comment (leandertolksdorf) [reply Apps/frontend/cypress/e2e/3-control-panel-top/gnd-button.cy.js:21]: As I didn't find a solution to this, I removed it just now and for next sprint we could consider using unit tests for such logic.

File: Apps/frontend/cypress/e2e/3-control-panel-top/general-buttons.cy.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+/// <reference types="cypress" />
+
+describe("GeneralButtons", () => {
+  beforeEach(() => {
+    cy.visit("http://localhost:5173/");
+  });
+  it("resets on off button when reset button is clicked", () => {
+    cy.get('[data-cy="on-off-button"]').click();
```
[the remark this answers is not in the document]
Comment (leandertolksdorf) [reply Apps/frontend/cypress/e2e/3-control-panel-top/general-buttons.cy.js:8]: Not sure what you meant, but to test if the reset button resets the on off button, I have to first toggle the on off button, right?

File: Apps/frontend/cypress/e2e/3-control-panel-top/reset-button.cy.js [hunk 5]
```diff
@@ -13,4 +13,9 @@ describe("resetButton", () => {
   it("is clickable", () => {
     cy.get('[data-cy="reset-button"]').click();
   });
+
```
[the remark this answers is not in the document]
Comment (leandertolksdorf) [reply Apps/frontend/cypress/e2e/3-control-panel-top/reset-button.cy.js:16]: I added blank lines manually for now.

**Review by motschel123 — APPROVED** [review]
All tests running green, LGTM

File: Apps/frontend/src/components/GNDButton.svelte (hunk 2 above)
[the remark this answers is not in the document]
Comment (PhlppKrmr) [reply Apps/frontend/src/components/GNDButton.svelte:25]: It reformats to multiline, but nothing major.

File: Apps/frontend/cypress/e2e/2-control-panel/preset-config.cy.js [hunk 6]
```diff
@@ -17,4 +22,10 @@ describe("everything visible on front page", () => {
     cy.get('[data-cy="selected-preset"]').should("be.visible");
     cy.get('[data-cy="store-channel-config"]').should("be.visible");
   });
+
+  it("closes popup on click outside", () => {
+    cy.get('[data-cy="settings-button"]').click();
```
[the remark this answers is not in the document]
Comment (PhlppKrmr) [reply Apps/frontend/cypress/e2e/2-control-panel/preset-config.cy.js:27]: As far as I can tell the interim step of checking if the panel is/was open is missing.
`cy.get('[data-cy="preset-config-open-popup"]').should("be.visible");`

File: Apps/frontend/cypress/e2e/3-control-panel-top/general-buttons.cy.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (PhlppKrmr) [reply Apps/frontend/cypress/e2e/3-control-panel-top/general-buttons.cy.js:9]: Yeah, I just meant functionality-wise you would not actually need the extra step. When considering the label of the test it is even expected that you also click the button. My misunderstaning

**Review by PhlppKrmr — APPROVED** [review]

**Pull request merged**